Sources/PromotedCore/Xray.swift

```diff
@@ -281,45 +334,108 @@ fileprivate extension Xray {
                          error.localizedDescription)
     guard let pendingBatch = pendingBatch else { return }
     pendingBatch.networkEndTime = clock.now
-    pendingBatch.error = error
+    pendingBatch.errors.append(error)
+    batchesWithErrors += 1
+    totalErrors += 1
   }
 
   private func batchResponseDidComplete() {
     osLog?.signpostEnd(name: "network")
     guard let pendingBatch = pendingBatch else { return }
     pendingBatch.networkEndTime = clock.now
     totalBytesSent += pendingBatch.messageSizeBytes
-    totalRequestsMade += 1
+    batchesSentSuccessfully += 1
     add(batch: pendingBatch)
     self.pendingBatch = nil
     logBatchResponseCompleteStats()
   }
 
   private func add(batch: NetworkBatch) {
-    networkBatches.append(batch)
-    trimNetworkBatches()
-  }
-
-  private func trimNetworkBatches() {
-    let excess = networkBatches.count - networkBatchWindowSize
-    if excess > 0 {
-      // networkBatches.removeFirst(k) is O(networkBatches.count).
-      // Not worth optimizing right now, but could use a circular
-      // buffer to avoid this.
-      networkBatches.removeFirst(excess)
-    }
+    networkBatchQueue.pushBack(batch)
   }
 
   private func logBatchResponseCompleteStats() {
-    guard let osLog = osLog else { return }
+    guard osLogLevel >= .info, let osLog = osLog else { return }
     if Self.timingMayBeInaccurate {
       osLog.info("WARNING: Timing may be inaccurate when running in debug or simulator.")
     }
     if let batch = networkBatches.last {
-      osLog.info("Latest batch: %{private}@", String(describing: batch))
+      if xrayLevel >= .callDetails && osLogLevel >= .debug {
+        logOperationSummaryTable(batch: batch)
+        logMessageSummaryTable(batch: batch)
+      } else {
+        osLog.info("Latest batch: %{private}@", String(describing: batch))
+      }
+    }
+    osLog.info("Total: Millis: %{public}lld, Bytes: %{public}lld, Batches: %{public}d",
+               totalTimeSpent.millis, totalBytesSent, batchesSentSuccessfully)
+  }
+
+  private func logOperationSummaryTable(batch: NetworkBatch) {
```

Comment:
side-comment - is it confusing when log means log to Promoted vs log to osLog?

Reply:
Changed this to consoleLog* to clarify.